Add deprecation whiltelist and raise errors on non-whitelisted deprecation warnings

millerdev opened the pull request:
FYI @dimagi/team-commcare-hq Developers may wish to `export CCHQ_WHITELISTED_WARNINGS=ignore` in development environments to hide whitelisted warnings.

:blowfish: Review by commit.

## Safety Assurance

### Safety story

The whitelist only applies when running tests or when `CCHQ_WHITELISTED_WARNINGS` (environment variable) is set. Some (tested) code has been changed to eliminate deprecated code paths.

### Automated test coverage

Yes.

### QA Plan

No QA.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

millerdev committed:
Drop PYTHONWARNINGS environment setting

Originally added in 71e49e848f9dfe0b7b82569bdbb2a98e7826a6f9, possibly worked on Python 2, doesn't seem to do anything on Python 3.

millerdev committed:
Ignore deprecated "property_match" in tests

"property_match" is a feature of userreports. It is fine that we have tests for a deprecated feature, but those tests should probably be updated to use something like `warnings.catch_warnings()` to ignore the warnings so they do not cause errors or other deprecation output. Updating tests that way is out of scope for the current effort. The whitelist item serves as a future reminder to address this issue.

It also appears to be used in some tests that were not obviously testing that feature alone. Maybe those usages should be changed to use the preferred "boolean_expression" instead?

See corehq/apps/userreports/filters/factory.py

millerdev committed:
Error on deprecation warning when running tests

millerdev committed:
inspect.getargspec() is deprecated

DeprecationWarning: inspect.getargspec() is deprecated since Python 3.0, use inspect.signature() or inspect.getfullargspec()

millerdev committed:
Fix zero-arg middleware deprecation

RemovedInDjango40Warning: Passing None for the middleware get_response argument is deprecated.

millerdev committed:
request.is_ajax() is deprecated

RemovedInDjango40Warning: request.is_ajax() is deprecated. See Django 3.1 release notes for more details about this deprecation.

millerdev committed:
Improve docstring format

millerdev committed:
Fix corehq.motech.fhir default app config

RemovedInDjango41Warning: 'corehq.motech.fhir' defines default_app_config = 'corehq.motech.fhir.apps.FHIRAppConfig'. Django now detects this configuration automatically. You can remove default_app_config.

millerdev committed:
Fix default app config warnings

RemovedInDjango41Warning: '...' defines default_app_config = '...'. However, Django's automatic detection did not find this configuration. You should move the default config class to the apps submodule of your application and, if this module defines several config classes, mark the default one with default = True.

See https://docs.djangoproject.com/en/4.0/ref/applications/#for-application-authors

millerdev committed:
Whitelist warnings triggered during test startup

There are still deprecation warnings triggered by tests that now cause test failures. They will be addressed in a future commit.

millerdev committed:
Ignore "TestResult has no addExpectedFailure method"

This warning originates inside of Python's `unittest` framework, and possibly has something to do with `nose`, although I didn't find any evidence of that. On first glance it was not clear what in our test code was triggering it or how to resolve it, but it does appear to cause the intended outcome (test failure is ignored).

dannyroberts commented on settings.py:
```diff
@@ -1480,7 +1480,6 @@ def _pkce_required(client_id):
     INSTALLED_APPS = INSTALLED_APPS + ('corehq.apps.mocha',)
     import warnings
     warnings.simplefilter('default')
-    os.environ['PYTHONWARNINGS'] = 'd'  # Show DeprecationWarning
```
As I was reviewing the python `warning` docs to brush up for reviewing this PR, I came across https://docs.python.org/3/library/warnings.html#overriding-the-default-filter (see the second example). There, this line comes with the comment `# Also affect subprocesses`, which implies to me that the original purpose of this line was to make this change (if changing to default is a change at all) affect subprocesses as well.

Just wanted to raise that because I remember we had a conversation where neither of us was clear on what this point of setting the environment variable here was.

dannyroberts reviewed (commented):
Looks good to me, I'll let you sort out the test failure and then I can do another round of review

millerdev committed:
Add environment variable for whitelisted warnings

See https://docs.python.org/3/using/cmdline.html#envvar-PYTHONWARNINGS

millerdev committed:
Whitelist new warnings since Django 3.2 upgrade

millerdev committed:
Fix UsersAppConfig location

The entry in settings.INSTALLED_APPS could not be changed to "corehq.apps.users.UsersAppConfig" because UsersMiddleware relies on the entry to equal or end with "users". Therefore the app config is being moved where Django expects it to be.

millerdev committed:
Remove redundant default_app_config variables

millerdev committed:
Augment deprecation warnings with source module

millerdev committed:
Witelist sqlalchemy warnings

Warnings triggered by corehq/apps/userreports/rebuild.py:get_table_diffs

Can be reproduced by commenting out these whitelist entries and running the test corehq.apps.userreports.tests.test_location_data_source:TestLocationDataSource.test_location_data_source

millerdev committed:
UnitTest.assert_() is deprecated

DeprecationWarning: Please use assertTrue instead.

millerdev committed:
tastypie: request.is_ajax() is deprecated

RemovedInDjango40Warning: request.is_ajax() is deprecated. See Django 3.1 release notes for more details about this deprecation.

millerdev committed:
json_response is deprecated

DeprecationWarning: json_response is deprecated.  Use django.http.JsonResponse instead.

Found in test: corehq.apps.app_manager.tests.test_views:TestViews.test_basic_app

millerdev committed:
Fix bool(XML(...)) warning

FutureWarning: The behavior of this method will change in future versions. Use specific 'len(elem)' or 'elem is not None' test instead.

millerdev committed:
Ignore "Overriding setting DATABASES" warning

millerdev committed:
Decimal years and months is deprecated

DeprecationWarning: an integer is required (got type decimal.Decimal).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.

millerdev committed:
Fix BeautifulSoup warning

GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 83 of the file corehq/apps/data_dictionary/tests/test_import_export.py. To get rid of this warning, pass the additional argument 'features="lxml"' to the BeautifulSoup constructor.

millerdev committed:
Fix SimplifiedUserInfo field access

The return type of `_report_user(user)` (was `_report_user_dict`) is `SimplifiedUserInfo`, which is an instance of `namedtupledict`. `namedtupledict.__getitem__()` and `get()` are deprecated.

DeprecationWarning: namedtuple fields should be accessed as attributes

millerdev committed:
inspect.getargspec() -> inspect.signature()

DeprecationWarning: inspect.getargspec() is deprecated since Python 3.0, use inspect.signature() or inspect.getfullargspec()

millerdev committed:
Filter motech.auth deprecation warning

DeprecationWarning: Non-string usernames will no longer be supported in Requests 3.0.0. Please convert the object you've passed in (None) to a string or bytes object in the near future to avoid problems.

millerdev committed:
Add @milldev as owner of corehq/warnings.py

millerdev committed:
Pass connection instead of engine

UserWarning: 'connection' argument to configure() is expected to be a sqlalchemy.engine.Connection instance, got Engine(***postgres:5432/test_commcarehq)

millerdev committed:
Logger.warn() is deprecated

Temporary whitelist until django-digest is updated with a fix.

DeprecationWarning: The 'warn' method is deprecated, use 'warning' instead

millerdev committed:
Relax partial index reflection warnings

millerdev committed:
assertEquals() is deprecated

DeprecationWarning: Please use assertEqual instead.

millerdev commented on corehq/motech/models.py:
```diff
@@ -153,6 +153,10 @@ def get_requests(
+        # FIXME this may construct an auth manager with a null username.
+        # This can result in an auth header with a username of "None",
+        # which is almost certainly not correct. Basic and Digest auth
+        # are affected, and possibly others.
```
@kaapstorm FYI, more context can be found in commit https://github.com/dimagi/commcare-hq/pull/31482/commits/e52ea875ab36c59b8f5a70d7219846da2e66605d. Is this something you could fix?

kaapstorm replied:
[OAuth1Manager](https://github.com/dimagi/commcare-hq/blob/99aa9dcf4e23bde6223f9f27906f8ab5a369dd2a/corehq/motech/auth.py#L125) and [OAuth2ClientGrantManager](https://github.com/dimagi/commcare-hq/blob/99aa9dcf4e23bde6223f9f27906f8ab5a369dd2a/corehq/motech/auth.py#L174) don't take a `username` parameter. For auth managers that do take `username`, we could assert that it has a value in the constructor. Does that sound right?

millerdev replied:
Possibly. I think the problem is that `ConnectionSettings` can exist in a state that should not be allowed (null username with incompatible auth type). One way to prevent this would be to add a [check constraint](https://www.postgresql.org/docs/9.4/ddl-constraints.html#DDL-CONSTRAINTS-CHECK-CONSTRAINTS) to either the `auth_type` or `username` field, disallowing a null value in `username` when `auth_type` is either `'basic'` or `'digest'` (and possibly other auth types that require a username). That way a `ConnectionSettings` row could not be saved in an invalid state.

A less correct option would be to `return AuthManager()` here if the auth type does not support a null username and `self.username is None`. This option does not require that all existing invalid data states be cleaned up, and will continue to allow invalid connection settings configurations to be saved in the database.

kaapstorm replied:
https://github.com/dimagi/commcare-hq/pull/31487

kaapstorm replied:
(Sorry, I wrote that before reading your comment. Kinda wish GitHub loaded comments without requiring a refresh. I agree that preventing invalid ConnectionSettings instances is better. I'll create a ticket for that, and follow up.)

millerdev committed:
Prevent filter_warnings() usage in non-test code

millerdev committed:
Document resolutions for deprecation errors

millerdev committed:
Do not use warnings whitelist unless testing...

Or when the environment is explicitly configured to do so. The whitelist has a cost, and it is better not to incur that cost in production environments. Developers may wish to `export CCHQ_WHITELISTED_WARNINGS=ignore` in development environments to hide whitelisted warnings.

millerdev committed:
Move comment

dannyroberts commented on corehq/warnings.py:
```diff
@@ -0,0 +1,174 @@
+        "^Skipped unsupported reflection of expression-based index form_processor_xformattachmentsql_blobmeta_key")
+    whitelist("sqlalchemy.", ignored_predicates, SAWarning)
+    whitelist("sqlalchemy.", skipped_reflections, SAWarning)
+    whitelist("unittest.case", "TestResult has no addExpectedFailure method", RuntimeWarning)
```
If possible/practical, I'd love to see this as a static list defined outside of this function, and then having this function simply iterate over that static list and whitelist each item. The benefit I see of that is that if we expect a large-ish group of developers who are only peripherally aware of the underlying plumbing to be adding and removing deprecation warnings from the list, it seems safer and more straightforward to ask them to edit a static list than to edit the contents of a function.

millerdev replied:
https://github.com/dimagi/commcare-hq/pull/31482/commits/3ef6d2f1620488b1936106672651ae86674d5742

millerdev committed:
Warnings whitelist as a static list of tuples
